### DISCUSSION TITLE
Random access

### DISCUSSION THREAD

**Discussion by michaelkirk** [flatgeobuf/flatgeobuf#291] [General]
> Goals are to be suitable for large volumes of static data, significantly faster than legacy formats without size limitations for contents or metainformation and to be suitable for streaming/random access.

I'm trying to clarify my understanding about what is meant by random access here.

Let's review the layout of an FGB file

![layout](https://github.com/flatgeobuf/flatgeobuf/blob/master/doc/layout.svg)

## layout
MB: Magic bytes (0x6667620366676201)
H: Header (variable size [flatbuffer](https://github.com/flatgeobuf/flatgeobuf/blob/master/src/fbs/header.fbs))
I (optional): Static packed Hilbert R-tree index (static size [custom buffer](https://github.com/flatgeobuf/flatgeobuf/blob/master/src/cpp/packedrtree.h))
DATA: Features (variable size [flatbuffer](https://github.com/flatgeobuf/flatgeobuf/blob/master/src/fbs/feature.fbs)s)

Flatbuffers, as a major design goal, support random access of fields without needing to parse the whole structure. So, given the Header buffer, we can dig into just the fields we're interested in without parsing the ones we're not interested in. Cool.

Similarly, given an individual Feature buffer, we can access the fields of the Feature without parsing all of them. Notably though, these fields are only the geometry and column schema. The actual property data (e.g. the corollary to "Dinagat Islands" from https://geojson.org) is a custom serialization which isn't randomly accessible. It must be linearly deserialized.

Further, note that "DATA: Features" is not a single flatbuffer, but rather a series of concatenated size prefixed flatbuffers, schematically like this:

**Data: Feature** `size_1: u32, [$size_1 bytes of feature data ], size_2: u32, [$size_2 bytes of feature data], ...`

Consequently, on it's own, "DATA: Features" must be accessed serially and cannot be accessed randomly. That is, if I want to get the 612th feature, I need to first read `size_1`, skip that many bytes, then read `size_2`, skip that many bytes, and so on all the way to `size_612` before I can read feature 612.

However, if the optional Index is present, we have the offsets of the individual features into the data buffer. That feels pretty close to random access. But we're primarily relying on the index (I), not on any quality of flatbuffers, to randomly access the Feature.

So a couple of observations:

1. Without the spatial index, accessing a random feature is expensive and bounded by the size of the feature collection.
2. With the spatial index, the cost of reading a Feature is bounded by the size of that single feature, it's not related to how many features you have.

I guess my realization was, flatbuffers aren't getting us as much as I thought here for our feature encoding. We're relying on the spatial index for random feature access, not flatbuffers. 

We do arguably get some benefits from flatbuffer - e.g. we can skip deserializing the properties data array if we're only interested in the geometry. I don't think I've gained much from having "random access" into the geometry, because typically, if you're reading the geometry we're reading *all* of the geometry, but I suppose some cases aren't interested in the geometry at all, in which case they'd benefit from not having parsed it.

And flatbuffers carries with it some baggage. The basic premise of having a "random access friendly" format like flatbuffers is to add padding (in other words, to waste space). Space saving approaches like varints or offset encodings are at odds with the format.

Sorry that this is a bit of a tirade. I still really like flatgeobuf and I think it's a big improvement. But I'm thinking about the future and want to make sure my understanding of the present is correct.

**Comment by bjornharrtell** [comment]
@michaelkirk all valid observations. :)

When I made flatgeobuf I was initially undecided if to use flatbuffers or protobuf. Initially I also used flatbuffers schema to represent properties but that turned out to be extremely space inefficient so I came up with the silly properties encoding instead. I think protobuf could perhaps have handled properties as general schema better but I'm not certain it would be space optimal. Also, it's worth noting that (also) protobuf isn't able to do much (any?) space saving on double array data.

Yes, random access requires the spatial index (which is also a feature offset index). That I ended up with spatial index being optional was actually an afterthought. My own use case for that was when using flatgeobuf as a serialization format to transfer a full set of data between two systems i.e a serial stream.

I'm still not sure if the decision to use flatbuffers was for the best. Mabye the popularity and space saving features of protobuf would have been a better trade off, I'm simply (still) not sure. I suppose I was persuaded by the "zero copy" properties of flatbuffers and that it would allow the coordinate data to be used directly from I/O.

Another thing/thought I had was that space saving should be a fully separate layer i.e compress with whatever you like, instead of being a format encoding/complexity thing.

Note that both flatbuffers and protobuf are designed for small "messages" so they are not well suited for things that can be very large. Both are using 32-bit internal addressing limiting a message to 2GB which I definitely didn't want to limit flatgeobuf to.

GeoArrow/GeoParquet is perhaps/probably a more thought through and optimized design and has the benefits of column oriented (and drawbacks?), however I think it's also far too complex in specification and implementation. Also it isn't stable yet and it looks like it will go into stable without https://github.com/opengeospatial/geoparquet/issues/13 so it can't really replace uses cases where spatial index is important in near future.

**Comment by michaelkirk** [comment]
Thanks for your response @bjornharrtell.

> Another thing/thought I had was that space saving should be a fully separate layer i.e compress with whatever you like, instead of being a format encoding/complexity thing.

This approach makes sense to me for things like repeated strings, where a generic compression pass could minimize things, but it's not obvious to me how to use a separate layer to achieve the kind of space savings for geometries you could get with var-ints or offset-encoded coordinates. Do you think it could work for that kind of thing?

> Note that both flatbuffers and protobuf are designed for small "messages" so they are not well suited for things that can be very large. Both are using 32-bit internal addressing limiting a message to 2GB which I definitely didn't want to limit flatgeobuf to.

Note that flatbuffers have recently added support for large messages: https://github.com/google/flatbuffers/pull/7935 (though I'm not sure if all the client languages have added support yet).

So maybe now all the Features could be stored in a single flatbuffer, but I'm not sure that we'd want to.

> GeoArrow/GeoParquet is perhaps/probably a more thought through and optimized design and has the benefits of column oriented (and drawbacks?), however I think it's also far too complex in specification and implementation. Also it isn't stable yet and it looks like it will go into stable without https://github.com/opengeospatial/geoparquet/issues/13 so it can't really replace uses cases where spatial index is important in near future.

Yes! I've been eyeing it from afar.

**Reply by bjornharrtell** [reply]
Agreed, compression would not do very much for raw double data. I did consider doing delta but wasn't sure it's possible to do 100% lossless with doubles which also was a design goal of mine. If you want to space optimal geobuf was a pretty neat design.

**Reply by bjornharrtell** [reply]
For reference, see https://github.com/mapbox/geobuf/issues/96 and note that they do not claim it to be lossless.

**Reply by michaelkirk** [reply]
That's a great point that I hadn't considered. Doing offset encodings probably implies some kind of fixed precision integer encoding, like OSM uses.

**Reply by bjornharrtell** [reply]
It was important for me make sure FlatGeobuf could roundtrip losslessly with other common double precision based storage.
